includes/class-send-list.php

```diff
@@ -0,0 +1,338 @@
+					'type'     => 'string',
+					'required' => true,
+					'enum'     => [
+						'list',
+						'sublist',
+					],
+				],
+				'entity_type' => [
+					'name'     => 'entity_type',
+					'type'     => 'string',
+					'required' => true,
+				],
+				'id'          => [
+					'name'     => 'id',
+					'type'     => 'string',
+					'required' => true,
+				],
+				'value'       => [
+					'name'     => 'value',
+					'type'     => 'string',
+					'required' => false,
+				],
+				'name'        => [
+					'name'     => 'name',
+					'type'     => 'string',
+					'required' => true,
+				],
+				'label'       => [
+					'name'     => 'label',
+					'type'     => 'string',
+					'required' => false,
+				],
+				'local_name'  => [
+					'name'     => 'local_name',
+					'type'     => 'string',
+					'required' => false,
+				],
+				'count'       => [
+					'name'     => 'count',
+					'type'     => 'integer',
+					'required' => false,
+				],
+				'parent_id'   => [
+					'name'     => 'parent_id',
+					'type'     => 'string',
+					'required' => false,
+				],
+				'edit_link'   => [
+					'name'     => 'edit_link',
+					'type'     => 'string',
+					'required' => false,
+				],
+			],
+		];
+	}
+
+	/**
+	 * Helper method to get a specific property's value.
+	 * Not for public use. Use specific property getters instead.
+	 *
+	 * @param string $key The property to get.
+	 *
+	 * @return mixed|WP_Error The property value or null if not set/not a supported property.
+	 */
+	private function get( $key ) {
+		$value  = $this->{ $key } ?? null;
+		$schema = $this->get_config_schema();
+		if ( ! empty( $schema['properties'][ $key ]['required'] ) && null === $this->{ $key } ) {
+			return new WP_Error( 'newspack_newsletters_send_list_missing_required_property_' . $key, __( 'Could not get required property: ', 'newspack-newsletters' ) . $key );
+		}
+
+		return $value;
+	}
+
+	/**
+	 * Helper method to set a property's value.
+	 * Not for public use. Use specific property setters instead.
+	 *
+	 * @param string $key The property to get.
+	 * @param mixed  $value The value to set.
+	 *
+	 * @return mixed|WP_Error The property value or WP_Error if not set/not a supported property.
+	 */
+	private function set( $key, $value ) {
+		$schema = $this->get_config_schema();
+		if ( ! isset( $schema['properties'][ $key ] ) ) {
+			return new WP_Error( 'newspack_newsletters_send_list_invalid_property_' . $key, __( 'Could not set invalid property: ', 'newspack-newsletters' ) . $key );
+		}
+
+		// If the passed value isn't in the enum, throw an error.
+		$property = $schema['properties'][ $key ];
+		if ( isset( $property['enum'] ) && ! in_array( $value, $property['enum'], true ) ) {
+			return new WP_Error( 'newspack_newsletters_send_list_invalid_property_value_' . $key, __( 'Invalid value for property: ', 'newspack-newsletters' ) . $key );
+		}
+
+		// Cast value to the expected type.
+		settype( $value, $property['type'] );
+
+		$this->{ $key } = $value;
```

Comment:
Setting the property directly into an undeclared object property has the downside of letting this prop public.

So I can still access `$send_list->label` which would skip our getter.

We could come up with some ideas to work around that, maybe using [overloading](https://www.php.net/manual/en/language.oop5.overloading.php) somehow, but I think the easiest thing to do is to store all properties inside one `private $config` property, which will be an array holding all the properties...

makes sense?

Reply:
Ah, I didn't realize that! 69e8232 refactors to use a single protected `$config` property instead of the separate undeclared properties. This also makes the `to_array()` method a bit simpler.